sonar-dotnet-shared-library/src/main/java/org/sonarsource/dotnet/shared/plugins/AbstractGlobalProtobufFileProcessor.java

```diff
@@ -78,17 +80,25 @@ private void processMetadataReportIfPresent(Path reportPath) {
       LOG.debug("Processing {}", metadataReportProtobuf);
       FileMetadataImporter fileMetadataImporter = new FileMetadataImporter();
       fileMetadataImporter.accept(metadataReportProtobuf);
-      this.generatedFileUris.addAll(fileMetadataImporter.getGeneratedFileUris());
-      this.roslynEncodingPerUri.putAll(fileMetadataImporter.getEncodingPerUri());
+      this.generatedFileUris.addAll(fileMetadataImporter.getGeneratedFileUris().stream().map(URI::toString).collect(Collectors.toList()));
+      for (Map.Entry<URI, Charset> entry : fileMetadataImporter.getEncodingPerUri().entrySet()) {
+        String key = entry.getKey().toString();
+        if (!this.roslynEncodingPerUri.containsKey(key)) {
+          this.roslynEncodingPerUri.put(key, entry.getValue());
+        } else if (this.roslynEncodingPerUri.get(key) != entry.getValue()) {
```

Comment:
could you do `this.roslynEncodingPerUri.get(key)` just once? store it in a var?

and maybe the same for `entry.getValue()` (you do the same for `entry.getKey()`)

Reply:
I considered that before and I've chosen it this way. This should be really rare error message, because the code path is almost unreachable (File.cs and file.cs with **different encoding** in the same project). The `TreeMap` has sublinear `O(log n)` asymptotic complexity so I'm fine with calling it twice in this context. And `entry.getValue()` does just `return this.value`.

I think it doesn't worth adding two more lines and one more level of nesting `if` to fix this.